cranelift/codegen/src/isa/x64/inst/mod.rs

```diff
@@ -404,6 +404,55 @@ pub enum Inst {
         offset: i64,
     },
 
+    // =====================================
+    // Instructions pertaining to atomic memory accesses.
+    /// A standard (native) `lock cmpxchg src, (amode)`, with register conventions:
+    ///
+    /// `dst`  (rd) address
+    /// `src`  (rd) replacement value
+    /// %rax   (mod) in: expected value, out: value that was actually at `dst`
+    /// %rflags is written.  Do not assume anything about it after the instruction.
+    ///
+    /// The instruction "succeeded" iff the lowest `ty` bits of %rax afterwards are the same as
+    /// they were before.
+    LockCmpxchg {
+        ty: Type, // I8, I16, I32 or I64
+        src: Reg,
+        dst: SyntheticAmode,
+        srcloc: Option<SourceLoc>,
+    },
+
+    /// A synthetic instruction, based on a loop around a native `lock cmpxchg` instruction.
+    /// This atomically modifies a value in memory and returns the old value.  The sequence
+    /// consists of an initial "normal" load from `dst`, followed by a loop which computes the
+    /// new value and tries to CAS (`lock cmpxchg{b,w,l,q}`) it into `dst`.  The loop iterates
+    /// until the CAS is successful.  If there is no contention, there will be only one pass
+    /// through the loop body.  The sequence does *not* perform any explicit memory fence
+    /// instructions (mfence/sfence/lfence).
+    ///
+    /// Note that the transaction is atomic in the sense that, as observed by some other thread,
+    /// `dst` either has the initial or final value, but no other.  It isn't atomic in the sense
+    /// of guaranteeing that no other thread writes to `dst` in between the initial load and the
+    /// CAS -- but that would cause the CAS to fail unless the other thread's last write before
+    /// the CAS wrote the same value that was already there.  In other words, this
+    /// implementation suffers (unavoidably) from the A-B-A problem.
+    ///
+    /// This insn has fixed register uses as follows:
+    ///
+    /// %r9   (rd) address
+    /// %r10  (rd) second operand for `op`
+    /// %r11  (wr) scratch reg; value afterwards has no meaning
+    /// %rax  (wr) the old value at %r9
+    /// %rflags is written.  Do not assume anything about it after the instruction.
+    AtomicRMW {
```

Comment:
I've tried to keep "Seq" in the name for synthetic sequences of instruction, can you put it as a suffix here please?